php/class-wp-https-detection.php

```diff
@@ -0,0 +1,140 @@
+<?php
+/**
+ * WP_HTTPS_Detection class.
+ *
+ * @package PWA
+ */
+
+/**
+ * WP_HTTPS_Detection class.
+ */
+class WP_HTTPS_Detection {
+
+	/**
+	 * The cron hook to check HTTPS support.
+	 *
+	 * @var string
+	 */
+	const CRON_HOOK = 'check_https_support';
+
+	/**
+	 * The option name for whether HTTPS is supported.
+	 *
+	 * @var string
+	 */
+	const HTTPS_SUPPORT_OPTION_NAME = 'is_https_supported';
+
+	/**
+	 * The query var key for HTTPS detection, used to verify that the request is from the same origin.
+	 *
+	 * @var string
+	 */
+	const REQUEST_QUERY_VAR = 'https_detection_token';
+
+	/**
+	 * The number passed to the query var.
+	 *
+	 * @var int
+	 */
+	public $query_var_number;
+
+	/**
+	 * Inits the class.
+	 */
+	public function init() {
+		add_action( 'wp', array( $this, 'schedule_cron' ) );
+		add_action( self::CRON_HOOK, array( $this, 'update_option_https_support' ) );
+		add_filter( 'cron_request', array( $this, 'ensure_http_if_sslverify' ), PHP_INT_MAX );
+		add_action( 'parse_query', array( $this, 'verify_https_check' ) );
+	}
+
+	/**
+	 * Schedules a cron event to check for HTTPS support.
+	 */
+	public function schedule_cron() {
+		if ( ! wp_next_scheduled( self::CRON_HOOK ) ) {
+			wp_schedule_event( time(), 'hourly', self::CRON_HOOK );
+		}
+	}
+
+	/**
+	 * Makes a request to find whether HTTPS is supported, and stores the result in an option.
+	 * If the request is a WP_Error, don't update the option.
+	 */
+	public function update_option_https_support() {
+		$https_support = $this->check_https_support();
+		if ( is_bool( $https_support ) ) {
+			update_option( self::HTTPS_SUPPORT_OPTION_NAME, $https_support );
+		}
+	}
+
+	/**
+	 * Makes a request to the home URL to determine whether HTTPS is supported.
+	 *
+	 * Returns true if the URL has the scheme of HTTPS, and it has the correct token.
+	 * The token is only to verify that the request is for the same site.
+	 * Also, in wp_remote_get(), 'sslverify' is true by default.
+	 *
+	 * @return boolean|WP_Error Whether HTTPS is supported, or a WP_Error.
+	 */
+	public function check_https_support() {
+		$this->query_var_number = wp_rand();
+		$response               = wp_remote_request(
+			add_query_arg(
+				self::REQUEST_QUERY_VAR,
+				$this->query_var_number,
+				home_url( '/', 'https' )
+			),
+			array(
+				'headers' => array(
+					'Cache-Control' => 'no-cache',
+				),
+			)
+		);
+
+		if ( is_wp_error( $response ) ) {
+			return $response;
+		}
+
+		if ( false === strpos( wp_remote_retrieve_body( $response ), wp_hash( $this->query_var_number, 'nonce' ) ) ) {
```

Comment:
This uses `strpos()` instead of a simple comparison because `wp_die()` wraps the `<body>` contents in some markup.

Reply:
Right before calling `wp_die()` you could do this:

```php
add_filter( 'wp_die_handler', function( $message ) {
    die( $message );
}, 1 );
```

This will then get overridden during unit tests to be `\WP_UnitTestCase::wp_die_handler()` via `\WP_UnitTestCase::get_wp_die_handler()`